File: 0865-smallest-subtree-with-all-the-deepest-nodes/0865-smallest-subtree-with-all-the-deepest-nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int height(TreeNode node) {
        //Base Case
        if(node==null) return 0;
        //Recusive call of faith
        int leftH = height(node.left);
        int rightH = height(node.right);
        //calculate my node's height 
        return Math.max(leftH,rightH) + 1;
    }
    public TreeNode subtreeWithAllDeepest(TreeNode root) {
        //base case
        if(root==null) return null;
        //get left and right rubtreeH
        int leftSubTreeH = height(root.left);
        int rightSubTreeH = height(root.right);
        
        if(leftSubTreeH==rightSubTreeH) return root;
        else if(leftSubTreeH<rightSubTreeH) return subtreeWithAllDeepest(root.right);
        else return subtreeWithAllDeepest(root.left);
    }
}

File: 0865-smallest-subtree-with-all-the-deepest-nodes/0865-smallest-subtree-with-all-the-deepest-nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int height(TreeNode* node) {
        //Base Case
        if(!node) return 0;
        
        //recursive faith of call of left and right height
        int leftH = height(node->left);
        int rightH = height(node->right);
        //Calculate meri height
        return max(leftH,rightH) +1;
    }
    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
        //Base case
        if(!root) return nullptr;
        //get max depth of left and right subtrees
        int leftSubTreeH = height(root->left);
        int rightSubTreeH = height(root->right);
        //check our cases conditions to conclude where we want to search for our subtree
        if(leftSubTreeH==rightSubTreeH) return root;
        else if(leftSubTreeH<rightSubTreeH) return subtreeWithAllDeepest(root->right);
        else return subtreeWithAllDeepest(root->left);
        
    }
};